lib/builder.ts

```diff
@@ -98,74 +104,67 @@ class Builder extends EventEmitter {
     }
   }
 
+  build() {
+    // return this.buildPromise();
+    if (this._cancelationRequest) {
+      return Promise.reject(new BuilderError('Cannot start a build if one is already running'));
+    }
+    let promise = this._buildAsync();
+    this._cancelationRequest = new CancelationRequest(promise);
+    return promise;
+  }
+
   // Trigger a (re)build.
   //
   // Returns a promise that resolves when the build has finished. If there is a
   // build error, the promise is rejected with a Builder.BuildError instance.
   // This method will never throw, and it will never be rejected with anything
   // other than a BuildError.
-  build() {
-    if (this._cancelationRequest) {
-      return Promise.reject(new BuilderError('Cannot start a build if one is already running'));
-    }
-
-    let promise = Promise.resolve();
-
+  private async _buildAsync() {
     this.buildId++;
 
+    // Wipe all buildState objects at the beginning of the build
     for (const nw of this._nodeWrappers.values()) {
-      // Wipe all buildState objects at the beginning of the build
       nw.buildState = {};
+    }
+
+    await forceAsync(); // run the remaining code in the next tick.
 
-      promise = promise.then(() => {
-        let needsEndNode = false;
-        // We use a nested .then/.catch so that the .catch can only catch errors
-        // from this node, but not from previous nodes.
-        return promiseFinally(
-          Promise.resolve()
-            .then(() => this._cancelationRequest.throwIfRequested())
-            .then(() => {
-              this.emit('beginNode', nw);
-              needsEndNode = true;
-            })
-            .then(() => nw.build()),
-          () => {
-            if (needsEndNode) {
-              this.emit('endNode', nw);
-            }
+    try {
+      // Build each node
+      for (const nw of this._nodeWrappers.values()) {
+        this._cancelationRequest && this._cancelationRequest.throwIfRequested()
```

Comment:
I believe this line should be contained within the next `try` so that the `Cancelation.isCancelationError(err)` bellow can catch it

Reply:
All that code does is catch it and re-throw it if it's a cancellation error. I don't see the point of moving this.

Reply:
After investigation this further it appears this may cause some additionally unexpected behavior as if `this._cancelationRequest` is not present the build can also not be cancelled. So guarding the existence of `this._cancelationRequest` suggests there is now a period where the build cannot be cancelled.